jedi/evaluate/docstrings.py

```diff
@@ -45,28 +49,147 @@
+        ...    'Parameters\n'
+        ...    '----------\n'
+        ...    'x : ndarray\n'
+        ...    'y : int or str or list\n'
+        ...    'z : {"foo", "bar", 100500}, optional\n'
+        ... )
+        >>> _search_param_in_numpydocstr(docstr, 'x')
+        ['ndarray']
+        >>> sorted(_search_param_in_numpydocstr(docstr, 'y'))
+        ['int', 'list', 'str']
+        >>> sorted(_search_param_in_numpydocstr(docstr, 'z'))
+        ['int', 'str']
+        """
         params = NumpyDocString(docstr)._parsed_data['Parameters']
         for p_name, p_type, p_descr in params:
             if p_name == param_str:
                 m = re.match('([^,]+(,[^,]+)*?)(,[ ]*optional)?$', p_type)
                 if m:
                     p_type = m.group(1)
-
-                if p_type.startswith('{'):
-                    types = set(type(x).__name__ for x in literal_eval(p_type))
-                    return list(types)
-                else:
-                    return [p_type]
+                return _expand_typestr(p_type)
         return []
 
+    def _search_return_in_numpydocstr(docstr):
+        r"""
+        Search `docstr` (in numpydoc format) for type(-s) of `param_str`.
+
+        >>> from jedi.evaluate.docstrings import *  # NOQA
+        >>> from jedi.evaluate.docstrings import _search_return_in_numpydocstr
+        >>> from jedi.evaluate.docstrings import _expand_typestr
+        >>> docstr = (
+        ...    'Returns\n'
+        ...    '----------\n'
+        ...    'int\n'
+        ...    '    can return an anoymous integer\n'
+        ...    'out : ndarray\n'
+        ...    '    can return a named value\n'
+        ... )
+        >>> _search_return_in_numpydocstr(docstr)
+        ['int', 'ndarray']
+        """
+        doc = NumpyDocString(docstr)
+        returns = doc._parsed_data['Returns']
+        returns += doc._parsed_data['Yields']
+        found = []
+        for p_name, p_type, p_descr in returns:
+            if not p_type:
+                p_type = p_name
+                p_name = ''
+
+            m = re.match('([^,]+(,[^,]+)*?)$', p_type)
+            if m:
+                p_type = m.group(1)
+            found.extend(_expand_typestr(p_type))
+        return found
+
+
+def _expand_typestr(p_type):
+    """
+    Attempts to interpret the possible types
+    """
+    # Check if alternative types are specified
+    if re.search('\\bor\\b', p_type):
+        types = [t.strip() for t in p_type.split('or')]
+    # Check if type has a set of valid literal values
+    elif p_type.startswith('{'):
+        if not is_py3:
+            # python2 does not support literal set evals
+            # workaround this by using lists instead
+            p_type = p_type.replace('{', '[').replace('}', ']')
+        types = set(type(x).__name__ for x in literal_eval(p_type))
+        types = list(types)
+    # Otherwise just return the typestr wrapped in a list
+    else:
+        types = [p_type]
+    return types
```

Comment:
This code was originally in `_search_param_in_numpydocstr` but I separated it out so I could re-use it. I also added the fix for the python2 `ast.literal_eval` problem here.